python/lsst/meas/algorithms/dynamicDetection.py

```diff
@@ -0,0 +1,210 @@
+            assert len(peaks) == 1
+            source.set(key, peaks[0].getF())
+            source.updateCoord(exposure.getWcs())
+
+        # Forced photometry on sky objects
+        self.skyMeasurement.run(catalog, exposure, catalog, exposure.getWcs())
+
+        # Calculate new threshold
+        fluxes = catalog["base_PsfFlux_flux"]
+        lq, uq = np.percentile(fluxes, [25.0, 75.0])
+        stdev = 0.741*(uq - lq)
+        errors = catalog["base_PsfFlux_fluxSigma"]
+        median = np.median(errors)
+        return median/stdev
+
+    def detectFootprints(self, exposure, doSmooth=True, sigma=None, clearMask=True):
+        """Detect footprints with a dynamic threshold
+
+        This varies from the vanilla ``detectFootprints`` method because we
+        do detection twice: one with a low threshold so that we can find
+        sky uncontaminated by objects, then one more with the new calculated
+        threshold.
+
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.Exposure`
+            Exposure to process; DETECTED{,_NEGATIVE} mask plane will be
+            set in-place.
+        doSmooth : `bool`
+            If True, smooth the image before detection using a Gaussian
+            of width ``sigma``.
+        sigma : `float`, optional
+            Gaussian Sigma of PSF (pixels); used for smoothing and to grow
+            detections; if `None` then measure the sigma of the PSF of the
+            ``exposure``.
+        clearMask : `bool`
+            Clear both DETECTED and DETECTED_NEGATIVE planes before running
+            detection.
+
+        Return Struct contents
+        ----------------------
+        positive : `lsst.afw.detection.FootprintSet`
+            Positive polarity footprints (may be `None`)
+        negative : `lsst.afw.detection.FootprintSet`
+            Negative polarity footprints (may be `None`)
+        numPos : `int`
+            Number of footprints in positive or 0 if detection polarity was
+            negative.
+        numNeg : `int`
+            Number of footprints in negative or 0 if detection polarity was
+            positive.
+        background : `lsst.afw.math.BackgroundMI`
+            Re-estimated background.  `None` if
+            ``reEstimateBackground==False``.
+        factor : `float`
+            Multiplication factor applied to the configured detection
+            threshold.
+        """
+        maskedImage = exposure.maskedImage
+
+        # We need a seed for the RNG in the sky object generator, but we don't have
+        # access to the butler, which is what we'd normally use to get a seed for a
+        # particular piece of data. We need something that changes with the input
+        # (so we don't use the same sky objects for every input), but deterministic.
+        # It also needs to fit in a C++ 'int' so pybind doesn't choke on it.
+        seed = int(maskedImage.image.array.sum()) % (2**31 - 1)
```

Comment:
Wouldn't that require changing the API of `SourceDetectionTask` as well, since it's supposed to be interchangeable? That seems a little strange to me.

Reply:
Yes, it would be required, and I think that's still the least-bad solution.  We may want to consider having a mechanism to *always* pass exposure IDs down through subtasks, or at least make it a very frequent optional argument even for Tasks that don't need it.  I suppose another option would be to add it to `Exposure` itself.  I don't think we want to do any of that on this ticket, of course - I just wanted to point out that this is a generic problem and we should resign ourselves to solving it (even if it's just a workaround now) rather than avoid it.

Reply:
I like the idea of attaching a `dataId` to `Exposure` via `VisitInfo` or similar.